### ISSUE TITLE
Further Parallelization Optimizations

### ISSUE THREAD

**Issue by rtfeldman** [rtfeldman/node-test-runner#193]
Originally I parallelized tests using a round-robin approach: when each worker process finishes running a test, it asks the supervisor which test to run next. When you get down to the final tests, this minimizes the amount of time cores are sitting idle waiting for the last one to finish.

However, in practice, the round-robin approach was much slower than having each of (say) 4 workers pick 1/4 of the tests and run those to completion. This is presumably faster because it eliminates the Inter-Process Communication overhead the round-robin approach requires. (If Node had threads that overhead would probably be negligible, but it doesn't, so here we are.)

There are still several things we could try, which might potentially speed this up even more. I'm not saying we should do all of these things, just writing them down as potential areas to explore.

### Parallelize Shrinking

When a test fails, shrinking takes longer than just about anything else. I haven't investigated to what extent shrinking could be parallelized, but if it can be, there are probably substantial wins there.

### Just-in-time round-robin

Start off by having all worker processes plow ahead without communicating with the supervisor, just like how they do today.

Once the first worker finishes, have it report to the supervisor "I'm finished, and now I'm idle." At this point we switch into round-robin model. The supervisor then tells all the other workers "finish up what you're doing, then report where you left off, and that you're idle." At that point the supervisor has all the information necessary to round-robin the remaining tests.

It's possible that this would avoid cases that are possible today, where one worker is holding up the show

### Parallelize fuzz test runs

By design, it's possible to have a fuzz test with n=100 run all 100 iterations in parallel.

We aren't currently doing this, but we could. I'd expect this to have higher memory overhead, but less CPU idling at the end of the test run, and thus an overall lower duration for the same test suite.

### Run slowest tests first

Do what Jest does, which is to sort tests by duration - I guess based on caching that from previous test runs, and/or possibly some heuristic - and run the slowest ones first.

For us, this is probably a matter of running fuzz tests first (especially in case they fail and require shrinking, although if we can parallelize shrinking and test-running in the first place, that might be less of a big deal), and someday Webdriver tests - which I would certainly expect to be the slowest.

It's also possible that users wouldn't want this, or perhaps would want to be able to turn it off. Certainly output can be printed in the original order regardless, but this could delay your getting info on what failed. (The thing is, usually if I care about order of test execution, I'm using something like `only` or `skip` anyway. So I suspect this may not matter.)

**Comment by gampleman** [comment]
Parallelizing fuzz tests would be great. In a rather extreme case I have a project with a single fuzz test that takes about 8 hours to complete (at a 1000 iterations) - parallelizing that would be a huge win.

Could you mix the round-robin and split in a way where you keep a little stash for the winner:
if you have say 4 workers, then give each worker 1/5 of the work, then round robin the remainder?

**Comment by rtfeldman** [comment]
@gampleman oh wow, that's great to know about!

That makes me think we should definitely attempt to parallelize individual fuzz tests.

Would you mind opening a new issue about your extreme case? (We can figure out implementation details later, e.g. the round robin, but it would be great to have the use case recorded in an issue!)

**Comment by gampleman** [comment]
Sure, what in particular would you like to know? I'll try to describe it a bit tomorrow.
